### PR TITLE
[FLINK-31632] Fix maxAllowedWatermark arithmetic overflow when the source is idle

### PR DESCRIPTION
**Pull request by haishui126**
## What is the purpose of the change

This PR aims to fix maxAllowedWatermark overflow when source is idle, which leads to source can't resume to active.

## Brief change log

check maxAllowedWatermark is overflow. if so, set Long.MAX_VALUE

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no) no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no) no
  - The serializers: (yes / no / don't know) no
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know) no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know) no
  - The S3 file system connector: (yes / no / don't know) no

## Documentation

  - Does this pull request introduce a new feature? (yes / no) no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Review by mas-chen — COMMENTED** [review]
The change makes sense. Can you add a unit test to verify the regression is fixed?

I also think there needs to be another change to the `shouldWaitForAlignment()` to not wait on max watermark

**Comment by haishui126** [comment]
I have added a test base on [FLINK-25982](https://issues.apache.org/jira/browse/FLINK-25982).
Is there anything else I need to do?

**Comment by haishui126** [comment]
> The change makes sense. Can you add a unit test to verify the regression is fixed?
> 
> I also think there needs to be another change to the `shouldWaitForAlignment()` to not wait on max watermark

After fixing the overflow, `currentMaxDesiredWatermark` = `lastEmittedWatermark` = Long.MAX_VALUE, the  `shouldWaitForAlignment()` will return false, and the idle source will keep READING. It looks like it won't need to be changed.

**Comment by aristofun** [comment]
Hi there. Thanks for your amazing work and fixing this. Any ETA when can it get into 1.16 branch?

File: flink-runtime/src/test/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorAlignmentTest.java
Comment (mas-chen) [inline flink-runtime/src/test/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorAlignmentTest.java:87]: Looks great!

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/SourceOperatorAlignmentTest.java
Comment (mas-chen) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/SourceOperatorAlignmentTest.java:170]: I think you want to check the `WatermarkAlignmentEvent`? and then assert on whether the maxWatermark is correct

**Comment by mas-chen** [comment]
@haishui126 can we add a more meaningful commit message such as "[FLINK-XYZ] Fix maxAllowedWatermark arithmetic overflow when the source is idle" or something similar?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/SourceOperatorAlignmentTest.java
Comment (haishui126) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/SourceOperatorAlignmentTest.java:170]: Yes! I have read the code just now, I find the `currentMaxDesiredWatermark` is maxWatermark by default, that's why the old test passed. I'm confused whether it's right to write like this now.

**Comment by haishui126** [comment]
> @haishui126 can we add a more meaningful commit message such as "[FLINK-XYZ] Fix maxAllowedWatermark arithmetic overflow when the source is idle" or something similar?

Absolutely! "[FLINK-31632] Fix maxAllowedWatermark arithmetic overflow when the source is idle" sounds great! Do we edit the commit message on merging this pr or I re-push?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/SourceOperatorAlignmentTest.java
Comment (mas-chen) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/SourceOperatorAlignmentTest.java:170]: Oh, I see. Does it make sense to call operator.handleOperatorEvent(new WatermarkAlignmentEvent(...)) after every `assertLatestReportedWatermarkEvent` (line 162)?

**Comment by mas-chen** [comment]
> > arithmetic overflow when the source is idle" or something similar?
> 
> Absolutely! "[[FLINK-31632](https://issues.apache.org/jira/browse/FLINK-31632)] Fix maxAllowedWatermark arithmetic overflow when the source is idle" sounds great! Do we edit the commit message on merging this pr or I re-push?

Let's try to do that before merge. You can squash the commits into 1 and amend the commit message. It will make it easier for the committer to help merge =)

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/SourceOperatorAlignmentTest.java
Comment (haishui126) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/SourceOperatorAlignmentTest.java:172]: I think so, and I have fixed it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java
Comment (pltbkd) [inline flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java:183]: nit: maybe we can check with `Long.MAX_VALUE-globalCombinedWatermark.getTimestamp() >= watermarkAlignmentParams.getMaxAllowedWatermarkDrift()`, instead of using the try-catch? 
Or maybe it's better to add a comment explaining why an ArithmeticException can be thrown here.

**Review by pltbkd — COMMENTED** [review]
Thanks for the contribution! 
I did some investigation to make sure that it's ok to use Watermark.MAX_WATERMARK.getTimestamp here. 
This PR overall LGTM, and I left only a minor comment.

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java
Comment (haishui126) [reply flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java:183]: Thanks for the suggestion, but the `Long.MAX_VALUE-globalCombinedWatermark.getTimestamp()` will overflow when `globalCombinedWatermark.getTimestamp()` is negated. If It's necessary to replace the try-catch, we can use: 
```java
        long maxAllowedWatermark =
                globalCombinedWatermark.getTimestamp()
                        + watermarkAlignmentParams.getMaxAllowedWatermarkDrift();
        // check maxAllowedWatermark arithmetic overflow when the source is idle
        if (((globalCombinedWatermark.getTimestamp() ^ maxAllowedWatermark)
                        & (watermarkAlignmentParams.getMaxAllowedWatermarkDrift()
                                ^ maxAllowedWatermark))
                < 0) {
            maxAllowedWatermark = Watermark.MAX_WATERMARK.getTimestamp();
        }
```
WDYT?

Comment (pltbkd) [reply flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java:183]: You are right, and I suppose `globalCombinedWatermark.getTimestamp() > 0 && Long.MAX_VALUE - globalCombinedWatermark.getTimestamp() < watermarkAlignmentParams.getMaxAllowedWatermarkDrift()` would be fine?
I think it better to be clear why and how it is checked. So in fact I would prefer simply adding comments in the try-catch to the bit operation if the simple expression is not enough.

Comment (haishui126) [reply flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java:183]: I add a comment to explain the ArithmeticException

Comment (pltbkd) [reply flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java:183]: Thanks for the update! LGTM now.

**Comment by MartijnVisser** [comment]
@mas-chen Are you OK with it? If so, @tzulitai do you want to have a final look before merging it?

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java
Comment (tzulitai) [inline flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java:187]: Could `ArithmenticException` be thrown due to any other reason outside of overflowing? If yes, it could be a bit dangerous to treat it like so.

Would it be sufficient to handle it like this?:
```
long maxAllowedWatermark = (globalCombinedWatermark.getTimestamp() != Watermark.MAX_WATERMARK.getTimestamp())
    ? globalCombinedWatermark.getTimestamp() + watermarkAlignmentParams.getMaxAllowedWatermarkDrift()
    : Watermark.MAX_WATERMARK.getTimestamp()
```

**Review by tzulitai — COMMENTED** [review]
Thanks for the PR @haishui126.

Had a minor comment about the overflow logic handling + excessive comments in tests.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/SourceOperatorAlignmentTest.java
Comment (haishui126) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/SourceOperatorAlignmentTest.java:173]: I agree with you. I‘ll edit the comment later.

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java
Comment (haishui126) [reply flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java:187]: This code solves the problem when a Source is idle.   I'm worried if there will be other cases that cause `globalCombinedWatermark.getTimestamp() + watermarkAlignmentParams.getMaxAllowedWatermarkDrift()` overflow.

Comment (tzulitai) [reply flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java:187]: Fair enough, I'm fine with leaving it as is then.

**Review by mas-chen — APPROVED** [review]

**Comment by tzulitai** [comment]
Merging ...

**Pull request closed without merging**
